Add SSL Certificate

piadelosreyes opened the pull request:
## Summary of changes

- _Add script to create SSL key and certificate_

/fyi: @tjoelsson, @mark-juwai : Please review.

Thanks

Commit:
[IND-376] Add SSL Certificate in redkeep

tjoelsson commented on tasks/ssl_certificate.yml:
```diff
@@ -0,0 +1,32 @@
+- name: create admin.juwai.com + digicert certificate file
+  template:
+    src: "admin.juwai.com.crt.j2"
+    dest: "/etc/ssl/admin.juwai.com.crt"
```
Use variable

tjoelsson commented on defaults/main.yml:
```diff
@@ -4,3 +4,7 @@ deploy_group: '{{ ansible_ssh_user }}'
 laravel_rotate_nginx_log: true
 laravel_rotate_nginx_log_retention: 30
+
+laravel_https_enabled: 'on'
```
Let's make this variable a boolean (yes/no) so adding certificate files and using https is optional. Not all applications will need it, but should still be able to use this role.

tjoelsson replied:
The default value should be no

tjoelsson commented on tasks/ssl_certificate.yml:
```diff
@@ -0,0 +1,32 @@
+    dest: "/etc/ssl/admin.juwai.com.key"
+    owner: root
+    group: '{{ web_server_user }}'
+    mode: 0640
```
We only want root to able to read this file. 640 is only for a special case where PHP needs to read the file too.

tjoelsson replied:
Also change the group to root.

tjoelsson commented on templates/admin.juwai.com.crt.j2:
```diff
@@ -0,0 +1 @@
+{{ admin_juwai_com_crt }}{{ digicert_crt }}
```
Name of this file should be generic

tjoelsson commented on templates/nginx.conf.j2:
```diff
@@ -17,6 +17,10 @@ server {
 
+    ssl                         {{ laravel_https_enabled }};
+    ssl_certificate             {{ laravel_https_certificate }};
+    ssl_certificate_key         {{ laravel_https_key }};
```
Wrap all these in an if-statement.

Commit:
Add SSL Certificate
- rename path into vars
- rename vars
- add nginx_use_ssl condition

tjoelsson commented on tasks/ssl_certificate.yml:
```diff
@@ -0,0 +1,32 @@
+    group: root
+    mode: 0640
+
+- name: create symlink so php saml library can find the key file
```
This symlink is not needed. It's only for SAML implementation.

tjoelsson commented on tasks/ssl_certificate.yml:
```diff
@@ -0,0 +1,32 @@
+    group: root
+    mode: 0640
+
+- name: create SSL certificate file
```
.sp file is also just for SAML. Can remove it.

tjoelsson commented on tasks/ssl_certificate.yml:
```diff
@@ -0,0 +1,32 @@
+    dest: "{{ laravel_https_key }}"
+    owner: root
+    group: root
+    mode: 0640
```
Change to 0600

Commit:
Removed symlink and cert source

tjoelsson commented on defaults/main.yml:
```diff
@@ -5,6 +5,6 @@ deploy_group: '{{ ansible_ssh_user }}'
 laravel_rotate_nginx_log_retention: 30
 
-laravel_https_enabled: no
+nginx_use_ssl: true
```
You can keep the name `laravel_https_enabled`. We want to have laravel_ as namespace for variables.

Commit:
Rename https enabled

tjoelsson commented on defaults/main.yml:
```diff
@@ -4,3 +4,7 @@ deploy_group: '{{ ansible_ssh_user }}'
 laravel_rotate_nginx_log: true
 laravel_rotate_nginx_log_retention: 30
+
+laravel_https_enabled: true
```
default to false

tjoelsson commented on templates/admin.juwai.com.crt.j2:
```diff
@@ -0,0 +1 @@
+{{ ssl_certs_crt }}{{ digicert_crt }}
```
Use a generic name for this file, like `ssl.crt.j2`

tjoelsson commented on tasks/ssl_certificate.yml:
```diff
@@ -0,0 +1,18 @@
+
+- name: create SSL + digicert certificate file
+  template:
+    src: "{{ cert_source }}"
```
This should be the template file

tjoelsson commented on tasks/main.yml:
```diff
@@ -88,6 +88,7 @@
     group: '{{ deploy_group }}'
   when: env == 'production' or env == 'staging'
 
+- include: ssl_certificate.yml
```
when laravel_https_enabled

tjoelsson reviewed (changes requested):
Add all new variables to readme

Commit:
Rename ssl certificate template file

- add new variables in readme
- set laravel_https_enabled  to default value False

Commit:
add condition to enable https in staging / prod

piadelosreyes commented:
Hi @tjoelsson, this is now ready for your review again. thanks

tjoelsson commented on tasks/main.yml:
```diff
@@ -89,7 +89,7 @@
 
 - include: ssl_certificate.yml
-  when: laravel_https_enabled
+  when: laravel_https_enabled and (env == 'production' or env == 'staging')
```
No need to check environment here. The variable can be set for different environments in playbook inventory files.

tjoelsson commented on templates/ssl.crt.j2:
```diff
@@ -0,0 +1 @@
+{{ ssl_certs_crt }}{{ digicert_crt }}
```
There are two more variables here. Need to document. Better use the same naming conventions for them as well.

These variables contain the actual certificate content. Maybe the path variables should end with _path.

Commit:
rename variable names

piadelosreyes commented:
@tjoelsson please review again. thanks.

tjoelsson commented on README.md:
```diff
@@ -77,6 +77,32 @@ Custom headers allowed by CORS, Access-Control-Allow-Headers.
+
+SSL certificate
+
+### laravel_https_digicert_data
```
Digicert is a company providing SSL certificates. Let's make this more generic and call it `laravel_https_chain_data`. Chain as in the certificate chain, which can include intermediate and root certificates.

piadelosreyes replied:
renamed in 05369c4

Commit:
Rename digicert variable to laravel_https_chain_data

tjoelsson commented on tasks/ssl_certificate.yml:
```diff
@@ -0,0 +1,18 @@
+---
+- include_vars: "group_vars/env-production/vault.yml"
+
+- name: create SSL + digicert certificate file
```
Just update this description and I think it's ok

Commit:
- update ssl description
